app/services/oauth_clients/one_drive_connection_manager.rb

```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2023 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module OAuthClients
+  class OneDriveConnectionManager < ConnectionManager
+    OAUTH_URI = URI('https://login.microsoftonline.com/')
+    DEFAULT_SCOPES = %w[offline_access files.readwrite.all user.read sites.readwrite.all].freeze
+
+    GRAPH_API_URI = URI('https://graph.microsoft.com')
+
+    def initialize(user:, oauth_client:, tenant_id:)
+      super(user:, oauth_client:)
+      @tenant_id = tenant_id
+    end
+
+    def authorization_state
+      current_token = get_existing_token
+      return :failed_authorization unless current_token
+
+      response = Net::HTTP.start(GRAPH_API_URI.host, GRAPH_API_URI.port, use_ssl: true) do |http|
+        http.get('/v1.0/me', { 'Authorization' => "Bearer #{current_token.access_token}" })
+      end
+
+      case response
+      when Net::HTTPSuccess
+        :connected
+      when Net::HTTPForbidden, Net::HTTPUnauthorized
+        service_result = refresh_token
+        if service_result.success?
+          :connected
+        elsif service_result.result == 'invalid_request'
+          :failed_authorization
+        else
+          :error
+        end
+      else
+        raise StandardError, 'not sure what to do'
+      end
+    rescue StandardError
+      :error
```

Comment:
I would suggest to just return error and not to handle all `StandardError`s. Let's just fail in unexpected situation or have a specific error class. Otherwise we will have no clue what is going on. Most likely it will be the error from line 64, but it is not 100%.
```suggestion
      :error
    end
```

Reply:
You have a point. I just copied the original implementation, that would need us to see what effects would this have on the FE side of the world.